libfwupdplugin/fu-plugin.c

```diff
@@ -1052,6 +1052,32 @@ fu_plugin_device_write_firmware (FuPlugin *self, FuDevice *device,
 	locker = fu_device_locker_new (device, error);
 	if (locker == NULL)
 		return FALSE;
+
+	/* back the old firmware up to /var/lib/fwupd */
+	if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_BACKUP_BEFORE_INSTALL)) {
+		g_autoptr(GBytes) fw_old = NULL;
+		g_autofree gchar *path = NULL;
+		g_autofree gchar *fn = NULL;
+		g_autofree gchar *localstatedir = NULL;
+
+		fw_old = fu_device_dump_firmware (device, error);
+		if (fw_old == NULL) {
+			g_prefix_error (error, "failed to backup old firmware: ");
+			return FALSE;
+		}
+		localstatedir = fu_common_get_path (FU_PATH_KIND_LOCALSTATEDIR_PKG);
+		fn = g_strdup_printf ("%s.bin", fu_device_get_version (device));
+		path = g_build_filename (localstatedir,
+					 "backup",
+					 fu_device_get_id (device),
+					 fu_device_get_serial (device) != NULL ?
+						fu_device_get_serial (device) :
+						"default",
+					 fn, NULL);
+		if (!fu_common_set_contents_bytes (path, fw_old, error))
+			return FALSE;
```

Comment:
One thought - what do you think about making a `latest.bin` symlink?  I guess it depends on how this backup would actually be used.

Reply:
I think it's a bit confusing tbh. Latest compared to what? Latest downgrade? Latest recovery?

Reply:
Most recent?  I want to be cognizant that version schema might be different between branches so latest is pretty confusing.  Maybe also need to include the branch in the path instead ?

Reply:
> version schema might be different between branches

Yes, you're right. We need both. Fixed.